chainer/optimizers/msvag.py

```diff
@@ -0,0 +1,178 @@
+from __future__ import division
+
+import numpy
+
+from chainer.backends import cuda
+from chainer import optimizer
+
+
+_default_hyperparam = optimizer.Hyperparameter()
+_default_hyperparam.lr = 0.1
+_default_hyperparam.beta = 0.9
+_default_hyperparam.eta = 1.0
+_default_hyperparam.weight_decay_rate = 0
+
+
+class MSVAGRule(optimizer.UpdateRule):
+
+    """Update rule of the M-SVAG optimization algorithm.
+
+    See: `Dissecting Adam: The Sign, Magnitude and Variance of Stochastic \
+          Gradients <https://arxiv.org/abs/1705.07774>`_
+
+    Modified for proper weight decay.
+
+    See: `Fixing Weight Decay Regularization in Adam \
+          <https://openreview.net/forum?id=rk6qdGgCZ>`_
+
+    See :class:`~chainer.optimizers.MSVAG` for the default values
+    of the hyperparameters.
+
+    Args:
+        parent_hyperparam (~chainer.optimizer.Hyperparameter): Hyperparameter
+            that provides the default values.
+        lr (float): Learning rate.
+        beta (float): Exponential decay rate of the first and second order
+                      moment.
+        eta (float): Schedule multiplier, can be used for warm restarts.
+        weight_decay_rate (float): Weight decay rate.
+
+    """
+
+    def __init__(self, parent_hyperparam=None,
+                 lr=None, beta=None,
+                 eta=None, weight_decay_rate=None):
+        super(MSVAGRule, self).__init__(
+            parent_hyperparam or _default_hyperparam)
+        if lr is not None:
+            self.hyperparam.lr = lr
+        if beta is not None:
+            self.hyperparam.beta = beta
+        if eta is not None:
+            self.hyperparam.eta = eta
+        if weight_decay_rate is not None:
+            self.hyperparam.weight_decay_rate = weight_decay_rate
+
+        self.beta_power = self.hyperparam.beta
+
+    def init_state(self, param):
+        xp = cuda.get_array_module(param.data)
+        with cuda.get_device_from_array(param.data):
+            self.state['m'] = xp.zeros_like(param.data)
+            self.state['v'] = xp.zeros_like(param.data)
+
+    def update_core_cpu(self, param):
+        grad = param.grad
+        if grad is None:
+            return
+        hp = self.hyperparam
+        m, v = self.state['m'], self.state['v']
+
+        rho = (((1.0 - hp.beta) ** 2) * (1.0 - self.beta_power ** 2) /
+               (((1.0 - self.beta_power) ** 2) * (1.0 - hp.beta ** 2)))
+        rho = min(rho, 0.99)
```

Comment:
The authors' implementation uses `0.9999`.  It seems to me that this line virtually forces `beta >= 0.01 / 2`.

Reply:
Good catch! Will change it.